### ISSUE TITLE
ExtrudedPolygon footprint uses nearest snapping: silently truncates polygon tips when bbox is not commensurate with the grid

### ISSUE THREAD

**Issue by em-junsun** [ymahlau/fdtdx#445]
## Summary

`_resolve_static_shapes` converts geometry-derived real shapes (e.g. the auto-inferred bbox of an `ExtrudedPolygon`) to grid-cell counts with nearest snapping (`round()` in v0.6.2; `snap="nearest"` on current main via `_real_length_to_grid_size`). When the bbox is not commensurate with the cell size, the allocated footprint can be **smaller than the polygon**, silently truncating polygon tips before `polygon_to_mask` ever runs — no warning is raised.

## Reproduction (v0.6.2, uniform grid)

Rotated rectangular pillar (ExtrudedPolygon) in a 400 nm unit cell:

| structure (wx,wy,θ) | bbox (nm) | n_alloc·Δ (nm) | truncation |
|---|---|---|---|
| (165,165,45°) | 233.34 | 23×10 = 230 | −3.35 nm (~1 tip cell) |
| (165,280,45°) | 314.66 | 31×10 = 310 | −4.66 nm (~1 tip cell) |
| (280,280,45°) | 395.98 | 79×5 = 395 | −0.98 nm |

Found in a 16-case rasterization audit that compared `place_objects` output cell-by-cell against an independent periodic reference rasterizer. Python banker's rounding applies, e.g. `round(23.334) = 23`.

Since `polygon_to_mask` only samples cell centers **inside the allocated window**, the truncated sliver is silently dropped. For rotated anisotropic meta-atoms this adds a geometry error of up to ~half a cell per side on top of staircasing, and it varies non-monotonically with resolution (different aliasing per Δ).

## Suggested fix

For geometry-derived shapes, snap **up** so the window always covers the polygon:

```python
shape_dict[obj_name][axis] = math.ceil(obj.partial_real_shape[axis] / config.resolution)
```

(or `snap="upper"` in `_real_length_to_grid_size` for geometry-derived inputs). Explicitly user-set `partial_grid_shape` / `partial_real_shape` could keep nearest snapping.

## Environment

fdtdx 0.6.2 (PyPI); main branch checked at 2026-09 (uniform-grid path in `_real_length_to_grid_size` still `snap="nearest"`); jax 0.11.0.

**Comment by ymahlau** [comment]
I was not able to reproduce this. Could you provide a more detailed description and possibly a minimal code working example?

**Comment by em-junsun** [comment]
Following up on my own report: after you could not reproduce this, I re-audited it at the mask level against an independent rasterizer, and **I'd like to retract the main claim** — sorry for the noise.

**Audit.** Independent cell-center rasterization (`matplotlib.path.Path.contains_points`) of the same vertices on the full, unwindowed grid, compared cell-by-cell against the material mask produced by `place_objects`:

```
(165,165,45deg) res=10nm | bbox=233.35nm exact_n=23.335 round=23 -> window 23 cells (230nm)
  fdtdx mask cells=265  independent-ref cells=265  LOST=0  GHOST=0

(173,173,45deg) res=10nm | bbox=244.66nm exact_n=24.466 round=24 -> window 24 cells (240nm)
  fdtdx mask cells=312  independent-ref cells=312  LOST=0  GHOST=0
```

(The second case is deliberately adversarial: `exact_n = 24.466` is as close to the 0.5-cell undershoot as a 45° square can get.)

**Why no cell is ever lost — structural, not a parameter coincidence.** `round()` undershoots by strictly less than 0.5 cells; `ExtrudedPolygon` is always centered in its allocated window; and the mask samples cell centers. Any cell center outside the window lies at least `(n/2 + 0.5)·Δ` from the center on some axis, while the polygon's bbox half-extent is strictly below `(n/2 + 0.25)·Δ` — so the polygon can never contain a cell center beyond the window, convex or not. Both window parities (odd/even cell counts) behave the same.

So the "truncation" column in my table was a window-vs-bbox size difference, **not** an observed mask difference, and the "silently dropped tips" wording was wrong. With cell-center sampling, nearest snapping and `ceil` produce identical masks — at most a cosmetic point.

The cell-by-cell differences in the internal audit that motivated this report most plausibly came from a different sampling convention in my reference rasterizer (coverage-based rather than cell-center); I should have verified the final semantic layer before filing. Thanks for pushing back — the repro request is exactly what forced the correction. Suggest closing as not-a-bug.

<details>
<summary>Audit script (fdtdx 0.6.2, CPU-only)</summary>

```python
import os
os.environ["JAX_PLATFORMS"] = "cpu"
import numpy as np
import jax, jax.numpy as jnp
import fdtdx
from fdtdx.objects.static_material.polygon import ExtrudedPolygon
from matplotlib.path import Path

NM = 1e-9; RES = 10 * NM; W = 165  # try W = 173 for the adversarial case
t = np.deg2rad(45); c, s = np.cos(t), np.sin(t)
v = 0.5 * NM * W * np.array([[-1, -1], [1, -1], [1, 1], [-1, 1]])
verts = v @ np.array([[c, -s], [s, c]]).T

poly = ExtrudedPolygon(name="pillar", axis=2, vertices=verts,
                       material_name="mat", materials={"mat": fdtdx.Material(permittivity=6.76)},
                       partial_grid_shape=(None, None, 32))
vol = fdtdx.SimulationVolume(partial_real_shape=(400 * NM, 400 * NM, 360 * NM))
bcfg = fdtdx.BoundaryConfig.from_uniform_bound(
    thickness=1, override_types={k: "periodic" for k in ("min_x", "max_x", "min_y", "max_y")})
bdict, clist = fdtdx.boundary_objects_from_config(bcfg, vol)
objects, _, _, _, _ = fdtdx.place_objects(
    object_list=[vol, poly] + list(bdict.values()),
    config=fdtdx.SimulationConfig(time=1e-14, resolution=RES, dtype=jnp.float32, courant_factor=0.99),
    constraints=list(clist) + [poly.place_at_center(vol, axes=(0, 1, 2))],
    key=jax.random.PRNGKey(0))
placed = next(o for o in objects.object_list if o.name == "pillar")
sl = placed.grid_slice
m = np.asarray(placed.get_voxel_mask_for_shape()); m2 = m[:, :, 0] if m.ndim == 3 else m
nx0, nx1, ny0, ny1 = sl[0].start, sl[0].stop, sl[1].start, sl[1].stop
vol_nx, vol_ny = objects.volume.grid_shape[:2]
cx, cy = 0.5 * (nx0 + nx1) * RES, 0.5 * (ny0 + ny1) * RES
gx = (np.arange(vol_nx) + 0.5) * RES; gy = (np.arange(vol_ny) + 0.5) * RES
pts = np.stack([np.repeat(gx, vol_ny), np.tile(gy, vol_nx)], axis=1)
ref = Path(verts + np.array([cx, cy])).contains_points(pts).reshape(vol_nx, vol_ny)
emb = np.zeros_like(ref); emb[nx0:nx1, ny0:ny1] = m2
print("fdtdx:", int(emb.sum()), " ref:", int(ref.sum()),
      " LOST:", int((ref & ~emb).sum()), " GHOST:", int((emb & ~ref).sum()))
```

</details>
